### ISSUE TITLE
Q: How to use the java dataloader

### ISSUE THREAD

**Issue by soudmaijer** [graphql-java-kickstart/graphql-java-tools#58]
I am looking at how to use the java dataloader: https://github.com/graphql-java/java-dataloader in my spring-boot-graphql app. Does graphql-java-tools provide support or does the spring-boot app provide this support?

I am trying to find some documentation to glue this all together.... any help would be really appreciated!

**Comment by apottere** [comment]
Right now the data loader isn't supported by this library or graphql-spring-boot, but there's been some interest and I'm looking at how I could add support for it.

**Comment by LeeWong** [comment]
It would be awesome. When do you think this feature can be ready? Thank you.

**Comment by apottere** [comment]
@LeeWong I'd have to look at how DataLoader is implemented in graphql-java, and I haven't had time to do that yet.  If you had experience I'd love to hear how you'd like to be able to use it with this library.

**Comment by szantogab** [comment]
@apottere I think we should have the ability to create new `GraphQL` object and instrumentations per request. I don't know what is the lifecycle of the `GraphQL` object, but [according to this doc](https://github.com/graphql-java/graphql-java/blob/master/docs/batching.rst#per-request-data-loaders) we should be able to specify that we want to have this object created in every request.

**Comment by soudmaijer** [comment]
@apottere As you already know dataloader is now included as a dependency for graphql-java v5.0.0. I am really curious what approach you are planning for integrating the dataloader. I was investigating a bit how this could/should work...

I create an experiment, see this gist: https://gist.github.com/soudmaijer/3fc6d32a2dad10960556a5e6ce0a5aca#file-graphqljavatoolsdataloader-kt-L21

What are your ideas?

**Comment by oexza** [comment]
Any update on this?

**Comment by oexza** [comment]
Any update on this?

**Comment by soudmaijer** [comment]
I have a working example for you.

Op 25 nov. 2017 09:47 schreef "Peace Michaels" <notifications@github.com>:

**Comment by apottere** [comment]
I don't have an update on this yet, working on subscription support then I'll focus on this.

**Comment by PaveLiArcH** [comment]
You can just provide `instrumentation` bean (if using graphql-spring-boot-starter)
```java
  @Bean
  DataLoaderRegistry dataLoaderRegistry(List<DataLoader<?, ?>> loaderList) {
    DataLoaderRegistry registry = new DataLoaderRegistry();
    for (DataLoader<?, ?> loader : loaderList) {
      registry.register(loader.getClass().getSimpleName(), loader);
    }
    return registry;
  }

  @Bean
  Instrumentation instrumentation(DataLoaderRegistry dataLoaderRegistry) {
    return new DataLoaderDispatcherInstrumentation(dataLoaderRegistry);
  }
```
create your DataLoaders, extending base DataLoader, and use DataLoaders in resolvers
```java
@Component
public class ProjectToVolumeDataLoader extends DataLoader<Integer, List<Volume>> {
  public ProjectToVolumeDataLoader() {
    this(null);
  }
  public ProjectToVolumeDataLoader(DataLoaderOptions options) {
    super(keys -> CompletableFuture.supplyAsync(() -> {
      List<List<Volume>> result = new ArrayList<>(keys.size());
      ... some magic to fill result (use db, use slow device, etc)  ...
      return result;
    }), options);
  }
}
```
```java
@Component
public class ProjectResolver implements GraphQLResolver<Project> {

  private final ProjectToVolumeDataLoader projectToVolumeDataLoader;

  public ProjectResolver(ProjectToVolumeDataLoader projectToVolumeDataLoader) {
    this.projectToVolumeDataLoader = projectToVolumeDataLoader;
  }

  public CompletableFuture<List<Volume>> getVolumes(Project project) {
    return projectToVolumeDataLoader.load(project.getProjectId());
  }
}
```

**Comment by tototoben** [comment]
Am I correct in assuming there isn't still a good way to have request scoped DataLoaders? I am using graphql-spring-boot and after some debugging through some resolvers I ended up feeling somewhat lost. 
Sorry if this is the wrong place to ask about this, I realize there is an [issue raised in the graphql-spring-boot repo (#44)](https://github.com/graphql-java/graphql-spring-boot/issues/44) that is of a similar nature, but it also seemed my concerns we're already raised and left open in this issue.

**Comment by freedomljc** [comment]
From my perspective, in the implementation of this project, the return type of resolving method is the actual value instead of the promise of the value, and it causes some problems when using the DataLoaderDispatcherInstrumentation: Normally the dataloader.dispatch() method will be invoked right after the resolving method. However, in the resolving method, considering it requires to return the actual value, it will cause the resolving method wait for the execution of dataloader.load() method, then indirectly wait for the execution of the dispatch method, then cause the thread hangs.

Without using the DataLoaderDispatcherInstrumentation, we can maintain the request-scope dataLoaderRegistry, and create dataloader when the request begins. Within the resolving method, invoke the dataloader.load() method, right after that, invoke dataloader.dispatch() method. But it will cause frequently invoking the dataloader.dispatch method.

**Comment by breun** [comment]
@tototoben I'm not sure what you consider 'a good way to have request scoped DataLoaders', but I'm successfully using DataLoader like this with graphql-spring-boot (manually translated from Kotlin to Java, so please forgive any syntax errors):

```
@Bean
public Instrumentation instrumentation(final MetricRegistry metricRegistry, final DataLoaderRegistry dataLoaderRegistry) {
    return DataLoaderDispatcherInstrumentation(dataLoaderRegistry);
}
            
@Bean
public DataLoaderRegistry dataLoadRegistry(final MyDataLoader myDataLoader) {
    DataLoaderRegistry dataLoaderRegistry = new DataLoaderRegistry();
    dataLoaderRegistry.register("myDataLoader", myDataLoader);
    return dataLoaderRegistry;
}
```

By the way, I had to manually override the version of java-dataloader to 2.0.2, because performance was pretty bad with 2.0.1, which comes with graphql-java 6.0, which comes with graphql-java-tools 4.3.0 (I believe because of a synchronisation issue in java-dataloader).

**Comment by breun** [comment]
Ah ok, we disable the DataLoader cache, because the default behaviour of the java-dataloader cache is indeed to cache without expiring (until application restart) and that's quite long. You can plug in your own cache implementation based on Guava or Kaffeine according to https://github.com/graphql-java/java-dataloader#custom-caches, but indeed it is not clear to me how to have a cache per request. So far I haven't had a need for caching (services called by our GraphQL applications may do their own caching), and since upgrading java-dataloader to 2.0.2 we can serve thousands of requests per second with just a couple of nodes.

**Comment by tototoben** [comment]
Looking at it now, yeah, just disabling the cache seems like a reasonable solution, thanks!

**Comment by yonigibbs** [comment]
Thanks @breun and @tototoben for the info above. I've disabled the cache as you suggest and it resolved the original problem described above, but it's exposed another small issue for us.  Not sure if we're doing something wrong, or if it's a genuine issue.  It's related to handling of duplicate data.  Say the request is for “user” objects, and their associated “company” objects.  And say you return users 1, 2, 3, who all belong to company 1.  If DataLoader caching is enabled, the “company” data loader is asked for company 1, and everything works as expected.  If the caching is disabled, what seems to happen in our tests is that the "company" data loader is asked for three companies (1, 1, 1).  The data access object returns a single company object, as expected, but an error then occurs:

_The size of the promised values MUST be the same size as the key list_

Obviously it’s fairly trivial to code around this, but that would then have to be done in every single data loader.  So I'm wondering whether disabling the cache is still the best solution.  Is there any other way to get request-scoped resolvers?  Or are we doing something wrong here?

**Comment by breun** [comment]
> If the caching is disabled, what seems to happen in our tests is that the "company" data loader is asked for three companies (1, 1, 1). The data access object returns a single company object, as expected

Why is returning a single company object expected? Shouldn't it return three company objects for the three company id's (even when they're all the same)?

**Comment by yonigibbs** [comment]
Thanks for the prompt answer @breun.

In our case we're just using a standard SQL data provider of some sort, e.g. JPA, which would end up running a query such as "SELECT * FROM Company WHERE ID IN (1,1,1)".  This gets back a single row, not three.  So we could code round it ourselves, but I'm just trying to avoid that if possible.

Also I'm just concerned in case completely turning off caching has other side effects and loses other useful functionality.  In the [DataLoader documentation](https://github.com/graphql-java/java-dataloader) it says "In certain **uncommon** cases, a DataLoader which does not cache may be desirable."  Have you come across any other side effects when turning off caching?

**Comment by breun** [comment]
Our applications work fine with caching disabled (but they're also not using an databases). I'm afraid you'll really need to change your code to return the same number of results as the number of inputs.

**Comment by yonigibbs** [comment]
Cool, thanks for the help!  In case anyone else comes across this, here's one possible way of doing this (in Kotlin), using an extension method:

```
// Extension method to act on an iterable of TValue objects, and return an iterable of (possibly null) values in the same order as the supplied keys.
// The passed in keyFetcher parameter is responsible for getting the key (of type TKey) from the object of type TValue.
fun <TValue, TKey> Iterable<TValue>.syncWithKeys(keyFetcher: (TValue) -> TKey, keys: Iterable<TKey>): Iterable<TValue?> {
    val objectsMap = this.associate({ Pair(keyFetcher(it), it) })
    return keys.map { key -> objectsMap[key] }.asIterable()
}

// Call it like this:
val companies = companyRepository.findAllById(keys)
return companies.syncWithKeys({ it.id }, keys)
```

**Comment by jarlehansen** [comment]
Does anyone have a complete example of using the java dataloader?
I am having a problem seeing the big picture here on how these parts are integrated.

**Comment by yonigibbs** [comment]
Hi @jarlehansen.  I don’t have an example I can share at the moment, though I’m working on it and will post a link if/when I get it done.  But in brief, here’s how I’ve managed to get it working, without having to turn off caching as explained above.  To give credit where it’s due, the solution to the caching question came mostly from @kaqqao's answer [here](https://stackoverflow.com/questions/47674981/how-to-resolve-graphql-n1-issue-with-graphql-jpa-java).

I’ll stick with the “user/company” example from above, i.e. you retrieve User objects from somewhere (e.g. a database), and then want to fetch related Company objects (e.g. also from a database).  So the GraphQL schema defines a User object with a property called “company” which returns an object of type Company.  But in your Java/Kotlin code, the object which wraps the user database record (e.g. “UserRecord”) has no “company” property: just a “companyId” property (e.g. a Long).

When you define your resolvers (which is done once, i.e. the resolvers are kinda singletons), don’t pass in the data loaders, as that’ll mean that these are singletons too, which is where the caching problem arises.  Instead, when handling the HTTP request create the loaders there (meaning that they are request-scoped) and store them in the execution input’s “context “and the GraphQL object’s “instrumentation”, as shown below:

```
val companyDataLoader = CompanyDataLoader(companyRepository)
val registry = DataLoaderRegistry()
registry.register("Company", companyDataLoader)

val executionInput = newExecutionInput()
        .query(graphQLParameters.query)
        .operationName(graphQLParameters.operationName)
        .variables(graphQLParameters.variables)
        .context(registry)

val graphQL = GraphQL
        .newGraphQL(schema)
        .instrumentation(DataLoaderDispatcherInstrumentation(registry))
        .build()
```

The CompanyDataLoader class is something like this:

```
class CompanyDataLoader(private val companyRepository: CompanyRepository) : DataLoader<Long, CompanyRecord>(BatchLoader { keys ->
    CompletableFuture.supplyAsync({
        // Fetch the companies, e.g. from a data repository
        companyRepository.findAllById(keys).toList()
    })
})
```

Next, in your UserResolver (which is used to resolve the “company” property on User objects), you’ll need to get this data loader and ask it to get the Company objects for you:

```
class UserResolver(private val companyRepository: CompanyRepository) : GraphQLResolver<UserRecord> {

    fun company(user: UserRecord, env: DataFetchingEnvironment): CompletableFuture<CompanyRecord> =
            env.getContext<DataLoaderRegistry>().getDataLoader<Long, CompanyRecord>("Company").load(user.companyId) 
}
```

The above approach provides per-request caching using DataLoader.

Additionally, there’s a very handy extension that you can use here, esp if working with SQL.  Typically, when fetching the user objects, you would run a query like “SELECT …  FROM User WHERE…”.  But, if the GraphQL request also asks for the users’ related company information, it would be most performant to do something like this: “SELECT …  FROM User INNER JOIN Company ON User.CompanyId = Company.Id WHERE…”.  This can be done as follows:

```
class Query(private val userRepository: UserRepository) : GraphQLQueryResolver {

    fun users(env: DataFetchingEnvironment): Iterable<UserRecord> {
        if (env.field.selectionSet.selections.any { it is Field && it.name == "company" }) {
            // GraphQL request contained the "company" field, so join to that table.
            // Run a query joining to that table and, for each user, get their company, and cache it in memory
            // in the company data loader.  DataLoader will then have this data in its cache (for this request)
            // so won't have to call the Company data loader to get the companies from the database.

            // First get the data loader from the data loader registry in the environment's context.
            val companyDataLoader = env.getContext<DataLoaderRegistry>().getDataLoader<Long, CompanyRecord>("Company")

            // Now loop round every record returned from the query and construct a CompanyRecord object from it,
            // then cache it as follows:
            companyDataLoader.prime(company.id, company)
        } else {
            // No request for company information: just query the User table directly with no joins
        }
    }
```

I hope this helps.  I'm totally new to GraphQL and DataLoader (and Kotlin!), so apologies if there's anything above I've misunderstood.  Please do correct my mistakes :-)

**Comment by kaqqao** [comment]
Yup, the scoping of the `DataLoaderRegistry` seems to confuse a lot of people. It should normally be instantiated per-request and it's the easiest to keep it in the shared GraphQL context param, as described above.

It is possible to make `DataLoaderRegistry` a managed bean and inject it using Spring BUT it is very inconvenient to use:

1. Make `DataLoaderRegistry` a request scoped bean
2. Make sure that when ever you run in a different thread, you propagate the thread-local `RequestAttributes` that Spring uses for request scoped beans. The easiest way to achieve this is to use a custom thread pool, similar in concept to `DelegatingSecurityContextExecutor ` used to propagate the thread-local `SecurityContextHolder` used for security.

That said, _please don't ever actually do that_, as it is incredibly easy to shoot yourself in the foot and you gain very little. Stick to the approach described in the previous comment instead.

**Comment by jarlehansen** [comment]
@yonigibbs thanks for the description, this is very useful.
We are using this with spring boot, do you know how I can use the ExecutionInput in this setup?

**Comment by jarlehansen** [comment]
@yonigibbs we are using the spring boot graphql starter, but I am thinking we might need to change this. Graphql webflux looks interesting, thanks for the tip. I will take a closer look at this.

Thank you so much for the assistance, I am new to graphql and without documentation it is hard to figure these things out. It would be very helpful to have a complete example, so please let me know when/if you get it ready.

**Comment by yonigibbs** [comment]
Here's a small sample project that uses WebFlux, GraphQL and DataLoader: https://github.com/yonigibbs/graphql-webflux-dataloader.  If you have any questions on it, feel free to create issues in that repository and hopefully I'll be able to answer.

**Comment by ashu-walmart** [comment]
Very new to GraphQl here, but seems to me that for Dataloaders to work with Spring MVC, changes may be needed in graphql-servlet.

1. Create an `InstrumenationFactory` and a default no-op implementation.
2. Add `Optional<Instrumentation>` to GraphQlContext
3. Update DefaultGraphQlContextBuilder to use InstrumentationFactory
4. Change SimpleGraphQlServlet.java to get instrumentation from GraphQlContext and set it in the GraphQl object created for each request.

The application may be able to make it work it by 
1. Defining an InstrumentationFactoryBean that returns DataLoaderRegistery either per request or from Spring managed beans as needed.
2. Define the DataLoaderDispatcherInstrumentation which uses InstrumentationFactory.

Does that make sense or am I way off here?

**Comment by jarlehansen** [comment]
Hi @ashu-walmart, the examples above written by @PaveLiArcH works. but as stated in the responses, there is a problem with sharing cache. We ended up using this method and disabling the cache functionality.

**Comment by kaqqao** [comment]
@jarlehansen I think @ashu-walmart 's approach is correctly scoping the `DataLoaderRegistery`, so no caching issues.
But honestly... I'd recommend replacing graphql-java-servlet with a simple Spring controller and calling it a day. The required functionality can likely be replicated in no time, and you retain full control over execution - so using DataLoader becomes a no-brainer instead of... well... just look at this thread.

**Comment by vojtapol** [comment]
I have successfully made the Java dataloader work with `graphql-java-tools` and Spring Boot. During startup I inject all repositories that can bulk fetch using Spring's list injection:
```
@Autowired
public RegistryBuilder(List<BatchLoader> bulkLoaders)
```
I use Spring's `ResolvableType` to get the generic type of the `BatchLoader` interface and put the loaders in a hash map. Then when creating the `DataLoaderRegistry` I just pull them from a map which saves some time. 

I create new `DataLoaderRegistry` for each request to avoid sharing the cache between requests. I distribute reference to the registry using the `DataFetchingEnvironment`. Pretty nifty.

### Problem
I do have one issue though. In a mutation I am doing some inserts and then when the returning type is resolving I use the data loader to fetch those rows that have just been inserted. Since they are not committed yet and data loader is dispatching on a different thread the uncommitted rows are not visible and the fetching fails.

I fixed this by adding `@Transactional(propagation = REQUIRES_NEW)` on the mutation method. This seems to commit the inserted rows before data loader starts fetching. Is this the correct approach? Any better ideas? Can I instead force the dispatcher to dispatch on one of Spring managed threads?

**Comment by ashu-walmart** [comment]
>I create new DataLoaderRegistry for each request to avoid sharing the cache between requests. I distribute reference to the registry using the DataFetchingEnvironment

Have you defined a custom Context to distribute to the resolvers using DataFetchingEnvironment? I am wondering if a change like this in graphql-servlet will make life easier: https://github.com/graphql-java/graphql-java-servlet/pull/83. As @kaqqao mentioned, otherwise the only option is to manually wire everything up as you have done.

**Comment by vojtapol** [comment]
I already had a custom `Context` for other purposes such as tracking authentication, accumulating errors or accessing user's locale. I just added this method:

```
public <T> CompletableFuture<T> load(@Nullable String id, Class<T> type) {
    if (id != null) {
        return this.registry.<String, T>getDataLoader(type.getCanonicalName()).load(id);
    }
    return null;
}
```

Which makes the use-site extremely clean:
```
return context.load(user.getAccountId(), Account.class);
```

I am not using `graphql-java-servlet` as it's not flexible enough.

**Comment by vojtapol** [comment]
Just a note of warning for whoever is implementing this.

It's not a good idea to do this:
```
class CompanyDataLoader(private val companyRepository: CompanyRepository) : DataLoader<Long, CompanyRecord>(BatchLoader { keys ->
    CompletableFuture.supplyAsync({
        // Fetch the companies, e.g. from a data repository
        companyRepository.findAllById(keys).toList()
    })
})
```
`supplyAsync` will by default use CommonForkJoin thread pool. That way part of your query might execute on the request thread but the data loader futures will be resolved on different threads. These are not not in the transaction opened by the request thread and will compete for database connections, eventually leading to a deadlock. The deadlock only occurs when you hit the server with more parallel request than the size of your database connection pool.

To prevent it do this instead:
```
supplyAsync(yourSupplier, new SyncTaskExecutor());
```

This way the futures will be resolved on the original request thread reusing the existing connection and running in the same transaction. I did not notice any significant slowdown by doing this.

Note that the above applies especially if you are recreating your data loaders for each request as described by @yonigibbs

**Comment by yonigibbs** [comment]
Thanks @vojtapol, that's really useful!  Can you please explain a bit more about why exactly the approach above is a problem?

Firstly, I presume that when you're talking about transactions you're not talking about normal database transactions (e.g. that are committed or rolled back), as we're just doing a SELECT here, not any INSERT or UPDATE or whatever.

Assuming that's the case... say a user makes a request to get some entity (say users in the example above), and this results in two data loaders being called (e.g. one for companies, and one for some other property related to users).  If a SyncTaskExecutor **isn't** used, then these two data loaders will run in parallel, and each will be on a different thread from the original request.  Each of these will then use a different database connection, as you say.  So is the issue then that, for a single user request, two database connections are being concurrently taken out the pool and used, meaning you'll run out of available connections more quickly?

If so, I have two questions:
1. Is there a way around this without forcing the two data loaders to run synchronously, which will slow things down? (Obviously this is all related to how exactly you get at data: we're using JOOQ, which uses JDBC, which is by nature blocking. Hence pushing stuff onto thread pools to get async behaviour which isn't strictly speaking truly non-blocking.)
2. What would actually happen if the database connections run out?  Would an exception be thrown, or would the threads that need to use the connections simply wait till one is available?

Thanks in advance!

**Comment by vojtapol** [comment]
@yonigibbs I am using Spring Boot with jOOQ so I think we have very similar set up. Note that I am not an expert but this is my understanding:

Let's fetch the User entity which has these fields: `id`, `name`, `companyId`, `accountId`. When resolving the incoming GraphQL query for list of users we load the user records from the database and then using the data loader we fetch companies and accounts in bulk. 

So when running data loaders on CommonForkJoin pool these queries are executed:

1. `getUsersByIds(ids)`  [request thread]
2. `getCompaniesByIds(ids)` [common-pool-1-thread-1]
3. `getAccountsByIds(ids)` [common-pool-1-thread-2]

Each query is running in a different thread and each will need it's own database connection. But note that only **after all the three queries are done** the request thread releases the database connection. (This is my assumption from what I have experienced)

Now imagine the situation of 200 clients hitting the API in parallel. This will happen:

1. 200 x `getUsersByIds(ids)`  200 x [request thread]

So all these request threads will quickly use up the connection pool and the rest will queue up which is as expected. But now the data loaders will try to load companies but will get stuck waiting for database connection because non of the request threads have released any connections yet. And that is the deadlock. Request threads can not finish the request and data loader threads can not get connections.  

To reproduce this issue try hitting your server with [ApacheBench](http://httpd.apache.org/docs/2.2/programs/ab.html) with `-c 200` and `-n 5000`. If your solution handles it fine, then my assumption about when connections are released is wrong and I have a different problem. My server would get stuck and then throw database connection time out errors.

**Comment by yonigibbs** [comment]
Thanks @vojtapol, that's a great explanation.  I've been doing some more investigations and experiments this morning, and below are my findings.  I'm no expert either, by the way: this is all quite new to me, so I might well be misunderstanding what's going on here.

First of all, you're totally right that using supplyAsync without specifying an executor (as I had in my original post) is not a good idea.  Thanks again for pointing that out: I wasn't aware of that.  This is explained in several places, e.g. [here](https://dzone.com/articles/which-thread-executes-completablefutures-tasks-and).

I tried to replicate your deadlock, and couldn't manage it.  Initially I left my code using the (bad) approach of not specifying an executor, and found that the CommonForkJoin pool that was created on my system had a pool size of 3 (this is all just running on my dev system, so by no means a realistic setup for load testing, but hopefully good enough for some rudimentary tests).  I added log messages to output what thread every query was getting executed on, and sure enough there were only ever three threads in action.  However even with such a tiny pool size, and with ApacheBench requesting 5000 requests, with a concurrency of 200, there was never a deadlock.  Obviously the responses were slow as there were 200 requests trying to be concurrently handled by three threads, but there were no errors.  So I **think** that the answer to my question above about what happens when an action is given to a thread pool in which all threads are already in use is that things get queued up, and will execute as soon as a thread is available.

I then updated the code to use an executor I'd explicitly created, giving it a larger pool size, and sure enough things got a bit faster, and again no errors.

So, I'm not sure what's different between your setup and mine.  Three possible things come to mind:

1. I'm using WebFlux.  Are you using this, or using MVC?
1. I noticed in your post above that you had "getUsersByIds(ids) [request thread]". So that sounds like you're running the initial query on the thread which made the request: is that right?  If so, we're doing things a bit differently: we're running **all** SQL on a thread pool (trying to get as close as possible to a truly non-blocking approach, within the limitations imposed by the fact that we're using JDBC).  So it's not only the data loaders that are run on the thread pool: it's also the initial query.  What this means is that the main thread pool which Spring Boot uses to serve the HTTP requests should never be getting used up (I hope!), and should always be able to handle more requests.
1. You said that you thought the DB connection wasn't getting returned till the end of the main request thread.  I'm not 100% sure how to test this, so not sure if that's what's happening on my system or not.  In case it's of use, here's how we're running each query:

```
        DriverManager
                .getConnection(url, username, password)
                .use { connection ->
                    DSL.using(connection, SQLDialect.POSTGRES_9_5).select().from(SOME_TABLE).fetch()
                }
```

I was under the impression that because we're using the [use](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/use.html) function in Kotlin, which ensures that the resource (in this case the DB connection) is closed down (which I guess means returned to the pool?), then the connection is returned to the pool as soon as the query has executed, not just when the whole request has been processed.

Anyway, that's what I've found so far.  Hope it's of use.  Please do correct anything I've misunderstood above.

One final thing which slightly puzzled me is this: as a test, I set my thread pool to a size of 15, but set jdbc.max-pool-size=5.  I was kind of expecting this to error, as I thought this would mean that 15 threads would be trying to concurrently use DB connections from a pool that was too small.  But I got no errors.  If anyone can explain that, I'd be grateful!

**Comment by ashu-walmart** [comment]
@vojtapol, Where are the transaction boundaries defined? Usually, I would create a Service Layer that has the transactional annotations and call that instead of calling the repository directly from the Resolvers.

**Comment by vojtapol** [comment]
@yonigibbs Thank you for your investigation. 

1. I am not using WebFlux, I use regular `@PostMapping` to handle our GraphQL request. This could definitely explain the differences.
2. Yes, the first JDBC fetch we do is on the request thread.
3. I am also not sure how to test it. I use `JooqAutoConfiguration` and then just inject `DSLContext` into repositories. I expect jOOQ to handle the connection to `DataSource` for us but I am not sure when it releases the connection. I think it tries to reuse the same connection until the request is fully resolved. But again, just guessing here. The `use` function in Kotlin should indeed close the connection right after the `fetch()` completes.

@ashu-walmart Right now I have only one `@Transactional` annotation defined on the GraphQL POST endpoint handler. The resolvers (before data loader) were just running in the same transaction boundaries. Maybe that's why I had to force them to run on the same thread for this to work?

**Comment by ashu-walmart** [comment]
@vojtapol Yes, I suspect that will cause other issue as well which will manifest as you start to scale your application out. I generally recommend to avoid wide long running transactions, but it really depends on your use case.

**Comment by vojtapol** [comment]
@ashu-walmart Thank you very much. You are very right. I limited the scope of my transactions and now there are no more deadlocks even when data loader uses multiple threads.

**Comment by kdlan** [comment]
I have tried to use dataloader in graphql-spring-boot-starter

https://gist.github.com/kdlan/30f5ac15859e0ed8204c5d41e65eae87

By adding `DataLoaderRegistry` to `HttpServletRequest.attribute` and passing it by `GraphQLContext`

```java
    @Bean
    GraphQLContextBuilder contextBuilder() {
        return (req, resp) -> {
            Context context = new StarWarsWiring.Context();
            req.get().setAttribute("context", context);
            req.get().setAttribute("__graphql_dataloader_registry", context.getDataLoaderRegistry());
            return new GraphQLContext(req, resp);
        };
    }
```

```java

    @Override
    public InstrumentationContext<ExecutionResult> beginExecution(InstrumentationExecutionParameters parameters) {
        DataloaderInstrumentationState state = parameters.getInstrumentationState();
        GraphQLContext context = parameters.getContext();
        state.init(log, context.getRequest().get());
        return super.beginExecution(parameters);
    }
```

**Comment by ashu-walmart** [comment]
@kdlan AFAIK to achieve request scoped data loading the instrumentation has to be set when initializing the GraphQL object per request. Is your use case similar and are you able to achieve similar request scoped data loading with your approach?

**Comment by ashu-walmart** [comment]
FWIW. This PR attempts to address that: https://github.com/graphql-java/graphql-java-servlet/pull/83

**Comment by kdlan** [comment]
@ashu-walmart Yes, The DataLoaderRegistry is created on each execution when  `GraphQLContextBuilder.build` called. And I rewrite the `DataLoaderDispatcherInstrumentation` to  init `FieldLevelTrackingApproach` and `InstrumentationState` at `beginExecution`,  to achieve request scope dataloader

**Comment by MatheusArleson** [comment]
Can someone share a example on how to use `@Batched` anotation?

If possible, one that uses just the library (no spring, react, ...).

**Comment by MatheusArleson** [comment]
Maybe this link will help : https://graphql-java.readthedocs.io/en/latest/batching.html

**Comment by MatheusArleson** [comment]
Thanks for the info, but from where you got this info?
No `@Deprecated` annotation or info on graphql-java master: [Batched annoation - v9.2](https://github.com/graphql-java/graphqlava/blob/master/src/main/java/graphql/execution/batched/Batched.java)

Can you share info about how its going to be the new impl?

**Comment by kaqqao** [comment]
The annotation itself is for some reason not marked as deprecated, but the only thing that makes it work is [BatchedExecutionStrategy](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/execution/batched/BatchedExecutionStrategy.java) which has been deprecated for a while now. This is because it's impossible to make it spec-compliant with regards to nulls and because it doesn't work with data loader instrumentation. In fact, _all_ execution strategies except the default `AsyncExecutionStrategy` are getting deprecated.

**Comment by vojtapol** [comment]
I have some more experience with it now. We are using Spring and if you write something during the request to database that is not committed yet, your data-loader-initiated db fetches will **not see the writes**. That's because they are executed on a different thread (likely one from `ForkJoinPool`) which does not have access to the same transaction.

I think the easiest way to fix it is to pass an instance of `SyncTaskExecutor` from request thread into the `supplyAsync` function that results in an instance of `DataLoader`. That way the loader will see all the uncommitted writes. Of course it will be slower for single requests but I think that under load if you saturate database connections the performance will be similar.

If you never have to read uncommited writes in your batched fetching, you can get away with parallel data loader execution.

**Comment by ccamel** [comment]
Here's my feedback. I managed to make the dataloaders work but with some tricks:
- like @vojtapol, I was faced a big issue regarding the management of the context in a multi-threaded environment. As a quick fix, I had to go to a synchronous approach in the `DataLoader` by using `CompletableFuture.completedFuture()` method call - instead of `CompletableFuture.supplyAsync()`.
- I also disabled the `DataLoader` cache as suggested by @breun since they're all singleton-scoped in my application (and not request-scoped as they should be).

**Comment by parthiNM** [comment]
> You can just provide `instrumentation` bean (if using graphql-spring-boot-starter)

>   public CompletableFuture<List<Volume>> getVolumes(Project project) {
>     return projectToVolumeDataLoader.load(project.getProjectId());
>   }

I'm wondering about from where to call the dispatch() or dispatchAndJoin()? I can see the Futures queued, but the results are not sent to the caller (PostMan).

Does anyone have a working Java example for graphql-spring-boot-starter and the schema in a .graphqls files?

**Comment by ccamel** [comment]
@parthiNM You don't have to call the `dispatch` methods by yourself. These methods are automatically called by the [DataLoaderDispatcherInstrumentation](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentation.java#L68) as each level of the query executes.

**Comment by dinhani** [comment]
`DataLoaderDispatcherInstrumentation` is not receiving a `DataLoaderRegistry` as a constructor parameter anymore since this breaking change:

https://github.com/graphql-java/graphql-java/pull/1263

**Comment by oliemansm** [comment]
@dinhani That breaking change is released in version 11.0 of graphql-java. The current version of graphql-java-tools supports graphql-java 10.0. Support for 11.0 has already been implemented but not released yet. I'm guessing you updated your own dependency of graphql-java to 11.0 already.

**Comment by oliemansm** [comment]
@dinhani `graphql-java-tools` will pull in the supported `graphql-java` dependency automatically, so if you simply remove it from your own dependencies you should be good.

**Comment by NikolaiChechin** [comment]
Hey! Does anybody have ideas how to use data loaders in a proper request-scoped way with `graphiql-spring-boot-starter`?

**Comment by oliemansm** [comment]
You mean `graphql-spring-boot-starter` I presume, not graphiql? Have you checked the examples in that other project you're referring too, because there is one: [example-request-scoped-dataloader](https://github.com/graphql-java-kickstart/graphql-spring-boot/tree/master/example-request-scoped-dataloader)

**Comment by NikolaiChechin** [comment]
@oliemansm Thanks a lot! That is exactly what I was looking for.

**Comment by bonobonohu** [comment]
> @parthiNM You don't have to call the `dispatch` methods by yourself. These methods are automatically called by the [DataLoaderDispatcherInstrumentation](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentation.java#L68) as each level of the query executes.
> 
> You should have such kind of configuration in your code:
> 
> ```java
> ...
> @Bean
> public Instrumentation instrumentation(DataLoaderRegistry dataLoaderRegistry) {
>   return new DataLoaderDispatcherInstrumentation(dataLoaderRegistry);
> }
> ```

bad news, it seems - at least for now - DataLoaderDispatcherInstrumentation has no constructor which takes a DataLoaderRegistry instance.
is this a problem, or the code was refactored properly? if it is a problem, are there any solutions?

thx!

**Comment by kaqqao** [comment]
@bonobonohu graphql-java has had this instrumentation registered by default for quite a while now, no need to add it yourself

**Comment by Silvatre** [comment]
Is there any update on how to add custom `DataLoader` instances into GraphQL ExecutionContext in `graphql-java-tools` generated schema/resolvers? There seems to be no simple bean to inject into the Spring context that would provide an ability to add DataLoaders. How is batching achieved on resolver layer with `graphql-java-tools` then?

**Comment by bonobonohu** [comment]
> @bonobonohu graphql-java has had this instrumentation registered by default for quite a while now, no need to add it yourself

okay, thanks, it seems i was on a completely bad track then! now i found this solution:
https://github.com/graphql-java-kickstart/graphql-spring-boot/tree/master/example-request-scoped-dataloader
and it just works like a charm. i refactored (back) my pure GrapQL-Java implementation back to utilize Kickstart, and experienced no issues (yet), but exactly the same behaviour.

**Comment by slavap** [comment]
@bonobonohu request scoped dataloader works for me as well. Just don't understand why you are saying that caching should be disabled? Per request scope eliminates caching problems, right?

**Comment by vojtapol** [comment]
You can pass data loader registry when building `ExecutionInput`:

```java
Builder executionInput = ExecutionInput.newExecutionInput()
    .dataLoaderRegistry(myRegistry)
    .query(query)
    .context(context);
```

This works as expected in GraphQL Java Tools. I would advise to make sure caching is either disabled or the whole registry is recreated for each request otherwise you might see invalid data being returned from the data loader.

Please open a new issue if you have more questions.

**Comment by Bwvolleyball** [comment]
@vojtapol - it's very hard to figure out what you mean with your reference to the `ExecutionInput` because there is no additional context.

**Comment by vojtapol** [comment]
@Bwvolleyball The `graphql.GraphQL` class has `.execute` methods. One of the overloads takes in an instance of `graphql.ExecutionInput`. That's what you see in my code snippet.

**Comment by soudmaijer** [comment]
For what it is worth, I pushed a full working example to my github repo. This is working with plain spring boot 2.2.5 (no graphql kickstart needed): 

https://github.com/sourcelabs-nl/graphql-java-tools-dataloader/blob/master/src/main/kotlin/nl/sourcelabs/graphql/demo/DemoApplication.kt

**Comment by dhitalsand** [comment]
> You can just provide `instrumentation` bean (if using graphql-spring-boot-starter)
> 
> ```java
>   @Bean
>   DataLoaderRegistry dataLoaderRegistry(List<DataLoader<?, ?>> loaderList) {
>     DataLoaderRegistry registry = new DataLoaderRegistry();
>     for (DataLoader<?, ?> loader : loaderList) {
>       registry.register(loader.getClass().getSimpleName(), loader);
>     }
>     return registry;
>   }
> 
>   @Bean
>   Instrumentation instrumentation(DataLoaderRegistry dataLoaderRegistry) {
>     return new DataLoaderDispatcherInstrumentation(dataLoaderRegistry);
>   }
> ```
> 
> create your DataLoaders, extending base DataLoader, and use DataLoaders in resolvers.
> Sample:
> 
> ```java
> @Component
> public class ProjectToVolumeDataLoader extends DataLoader<Integer, List<Volume>> {
>   public ProjectToVolumeDataLoader() {
>     this(null);
>   }
>   public ProjectToVolumeDataLoader(DataLoaderOptions options) {
>     super(keys -> CompletableFuture.supplyAsync(() -> {
>       List<List<Volume>> result = new ArrayList<>(keys.size());
>       ... do some magic to fill result (using db, using slow device, etc)  ...
>       ... result must be in same order as keys ...
>       return result;
>     }), options);
>   }
> }
> ```
> 
> ```java
> @Component
> public class ProjectResolver implements GraphQLResolver<Project> {
> 
>   private final ProjectToVolumeDataLoader projectToVolumeDataLoader;
> 
>   public ProjectResolver(ProjectToVolumeDataLoader projectToVolumeDataLoader) {
>     this.projectToVolumeDataLoader = projectToVolumeDataLoader;
>   }
> 
>   public CompletableFuture<List<Volume>> getVolumes(Project project) {
>     return projectToVolumeDataLoader.load(project.getProjectId());
>   }
> }
> ```

Instrumentation bean does not accept dataLoadRegistry..Any Alternatives ?
